### PR TITLE
Convert to Keras 3: `Knowledge Distillation` example

### PR DESCRIPTION
**Pull request by awsaf49**
This PR will add `Knowledge Distillation` example from `keras.io/examples`. Sadly, this example is not backend agnostic, presumably due to `tf.GradientTape()` which is TensorFlow specific.

### CODE REVIEW

**Comment by fchollet** [comment]
Formatted diff:

```diff
--- a/examples/keras_io/tensorflow/vision/knowledge_distillation.py
+++ b/examples/keras_io/tensorflow/vision/knowledge_distillation.py
@@ -1,12 +1,11 @@
 """
 Title: Knowledge Distillation
 Author: [Kenneth Borup](https://twitter.com/Kennethborup)
+Converted to Keras 3: [Md Awsafur Rahman](https://awsaf49.github.io)
 Date created: 2020/09/01
 Last modified: 2020/09/01
 Description: Implementation of classical Knowledge Distillation.
-Accelerator: GPU
 """
-
 """
 ## Introduction to Knowledge Distillation
 
@@ -29,12 +28,16 @@
 ## Setup
 """
 
+import os
+
+os.environ["KERAS_BACKEND"] = "tensorflow"
+
+import keras
+from keras import layers
+from keras import ops
 import tensorflow as tf
-from tensorflow import keras
-from tensorflow.keras import layers
 import numpy as np
 
-
 """
 ## Construct `Distiller()` class
 
@@ -50,8 +53,10 @@
 - An optimizer for the student and (optional) metrics to evaluate performance
 
 In the `train_step` method, we perform a forward pass of both the teacher and student,
-calculate the loss with weighting of the `student_loss` and `distillation_loss` by `alpha` and
-`1 - alpha`, respectively, and perform the backward pass. Note: only the student weights are updated,
+calculate the loss with weighting of the `student_loss` and `distillation_loss` by
+`alpha` and
+`1 - alpha`, respectively, and perform the backward pass. Note: only the student weights
+are updated,
 and therefore we only calculate the gradients for the student weights.
 
 In the `test_step` method, we evaluate the student model on the provided dataset.
@@ -111,8 +116,8 @@ def train_step(self, data):
             # as 1/T^2, multiply them by T^2 when using both hard and soft targets.
             distillation_loss = (
                 self.distillation_loss_fn(
-                    tf.nn.softmax(teacher_predictions / self.temperature, axis=1),
-                    tf.nn.softmax(student_predictions / self.temperature, axis=1),
+                    ops.softmax(teacher_predictions / self.temperature, axis=1),
+                    ops.softmax(student_predictions / self.temperature, axis=1),
                 )
                 * self.temperature**2
             )
@@ -168,7 +173,7 @@ def test_step(self, data):
     [
         keras.Input(shape=(28, 28, 1)),
         layers.Conv2D(256, (3, 3), strides=(2, 2), padding="same"),
-        layers.LeakyReLU(alpha=0.2),
+        layers.LeakyReLU(negative_slope=0.2),
         layers.MaxPooling2D(pool_size=(2, 2), strides=(1, 1), padding="same"),
         layers.Conv2D(512, (3, 3), strides=(2, 2), padding="same"),
         layers.Flatten(),
@@ -182,7 +187,7 @@ def test_step(self, data):
     [
         keras.Input(shape=(28, 28, 1)),
         layers.Conv2D(16, (3, 3), strides=(2, 2), padding="same"),
-        layers.LeakyReLU(alpha=0.2),
+        layers.LeakyReLU(negative_slope=0.2),
         layers.MaxPooling2D(pool_size=(2, 2), strides=(1, 1), padding="same"),
         layers.Conv2D(32, (3, 3), strides=(2, 2), padding="same"),
         layers.Flatten(),
@@ -198,7 +203,8 @@ def test_step(self, data):
 ## Prepare the dataset
 
 The dataset used for training the teacher and distilling the teacher is
-[MNIST](https://keras.io/api/datasets/mnist/), and the procedure would be equivalent for any other
+[MNIST](https://keras.io/api/datasets/mnist/), and the procedure would be equivalent for
+any other
 dataset, e.g. [CIFAR-10](https://keras.io/api/datasets/cifar10/), with a suitable choice
 of models. Both the student and teacher are trained on the training set and evaluated on
 the test set.
@@ -284,4 +290,4 @@ def test_step(self, data):
 You should expect the teacher to have accuracy around 97.6%, the student trained from
 scratch should be around 97.6%, and the distilled student should be around 98.1%. Remove
 or try out different seeds to use different weight initializations.
-"""
\ No newline at end of file
```

File: examples/keras_io/tensorflow/vision/knowledge_distillation.py [hunk 1]
```diff
@@ -0,0 +1,293 @@
+"""
+Title: Knowledge Distillation
+Author: [Kenneth Borup](https://twitter.com/Kennethborup)
+Converted to Keras 3: [Md Awsafur Rahman](https://awsaf49.github.io)
```
Comment (fchollet) [inline examples/keras_io/tensorflow/vision/knowledge_distillation.py:4]: Add: by

File: examples/keras_io/tensorflow/vision/knowledge_distillation.py [hunk 2]
```diff
@@ -0,0 +1,293 @@
...
+Date created: 2020/09/01
+Last modified: 2020/09/01
+Description: Implementation of classical Knowledge Distillation.
```
Comment (fchollet) [inline examples/keras_io/tensorflow/vision/knowledge_distillation.py:7]: Keep the Accelerator line

File: examples/keras_io/tensorflow/vision/knowledge_distillation.py [hunk 3]
```diff
@@ -0,0 +1,293 @@
...
+        self.temperature = temperature
+
+    def train_step(self, data):
+        # Unpack data
+        x, y = data
+
+        # Forward pass of teacher
+        teacher_predictions = self.teacher(x, training=False)
+
+        with tf.GradientTape() as tape:
```
Comment (fchollet) [inline examples/keras_io/tensorflow/vision/knowledge_distillation.py:107]: This is the only use of `tf` in the example. That means it's pretty close to being able to run on any framework.

The only train_step customization here is the computation of a custom loss. I believe it may be implemented in a fully backend-agnostic way by overriding `.compute_loss()`, which is a method shared across all backends.

The forward pass (`call()`) can be redirected to `student.call`, and `student_predictions` will be computed outside of `compute_loss()`. The signature is `compute_loss(self, x=None, y=None, y_pred=None, **kwargs)` -- x is the input, y_pred will be student_predictions.

**Review by fchollet — COMMENTED** [review]
Thanks for the PR!

File: examples/keras_io/tensorflow/vision/knowledge_distillation.py (hunk 3 above)
Comment (awsaf49) [reply examples/keras_io/tensorflow/vision/knowledge_distillation.py]: Thanks for tip. I've made the adjustments and now it is backend agnostic . I tried with `jax` and ran smoothly. 

Btw noticed a weird thing, `compute_loss(self, x=None, y=None, y_pred=None, **kwargs)` throws error with `jax` saying `number of arguments mismatch (takes 4 but 5 given)`. However, setting `compute_loss(self, x=None, y=None, y_pred=None, sample_weight=None, allow_empty=False)` solved the issue

Comment (awsaf49) [reply examples/keras_io/tensorflow/vision/knowledge_distillation.py]: Also as code is backend agnostic now, should I move the code to `./vision` from `./tensorflow/vision`?

File: examples/keras_io/tensorflow/vision/knowledge_distillation.py (hunk 2 above)
Comment (awsaf49) [reply examples/keras_io/tensorflow/vision/knowledge_distillation.py:7]: Added it back

File: examples/keras_io/tensorflow/vision/knowledge_distillation.py (hunk 3 above)
Comment (fchollet) [reply examples/keras_io/tensorflow/vision/knowledge_distillation.py]: I guess compiled JAX functions don't handle `kwargs` well. Better to keep the full signature then.

Did you try with torch as well?

Yes, the example can be moved out of the `tensorflow` folder now.

Comment (awsaf49) [reply examples/keras_io/tensorflow/vision/knowledge_distillation.py]: I'll move it to `./vision` then. Meanwhile, I'll also check if it works with `torch`

Comment (awsaf49) [reply examples/keras_io/tensorflow/vision/knowledge_distillation.py]: @fchollet just checked, works fine with `torch` even with `**kwargs`.

Comment (awsaf49) [reply examples/keras_io/tensorflow/vision/knowledge_distillation.py]: I have also moved the code to `/vision`

**Review by fchollet — APPROVED** [review]
Very nice!
